cmake/helpers/GdalGenerateConfig.cmake

```diff
@@ -0,0 +1,202 @@
+            if(_type STREQUAL "INTERFACE_LIBRARY")
+                continue()
+            endif()
+            gdal_get_target_property_for_config(_location "${_lib}" "IMPORTED_LOCATION")
+            if(_location)
+                set(_lib "${_location}")
+            endif()
+        endif()
+        set(_other "")
+        if(CMAKE_LIBRARY_PATH_FLAG STREQUAL "-L" AND _lib MATCHES "^-L")
+            set(_other "${_lib}")
+            set(_lib "")
+        elseif(_lib STREQUAL "-pthread")
+            # use _lib
+        elseif(CMAKE_LINK_LIBRARY_FLAG STREQUAL "-l" AND _lib MATCHES "^-l")
+            # use _lib
+        elseif(EXISTS "${_lib}")
+            gdal_split_library_to_lflags(_lib _other "${_lib}")
+        else()
+            set(_lib "${CMAKE_LINK_LIBRARY_FLAG}${_lib}")
+        endif()
+        if(_other AND NOT _other IN_LIST _libs_out)
+            list(APPEND _libs_out "${_other}")
+        endif()
+        if(_lib)
+            list(REMOVE_ITEM _libs_out "${_lib}")
+            list(APPEND _libs_out "${_lib}")
+        endif()
+    endwhile()
+    string(REPLACE ";" " " _libs_out "${_libs_out}")
+    set(${_result} "${_libs_out}" PARENT_SCOPE)
+endfunction()
+
+# Generate gdal-config utility command and pkg-config module gdal.pc
+function(gdal_generate_config)
+    set(one_value_keywords "TARGET;GLOBAL_PROPERTY;GDAL_CONFIG;PKG_CONFIG")
+    cmake_parse_arguments(PARSE_ARGV 0 arg "" "${one_value_keywords}" "")
+
+    if (ENABLE_GNM)
+        set(CONFIG_GNM_ENABLED "yes")
+    else ()
+        set(CONFIG_GNM_ENABLED "no")
+    endif ()
+    get_property(gdal_formats GLOBAL PROPERTY GDAL_FORMATS)
+    get_property(ogr_formats GLOBAL PROPERTY OGR_FORMATS)
+    string(REPLACE ";" " " CONFIG_FORMATS "${gdal_formats} ${ogr_formats}")
+
+    if(NOT DEFINED CMAKE_INSTALL_PREFIX)
+        set(CONFIG_PREFIX "/usr/local") # default
+    else()
+        set(CONFIG_PREFIX "${CMAKE_INSTALL_PREFIX}")
+    endif()
+    set(CONFIG_DATA "${CONFIG_PREFIX}/${GDAL_RESOURCE_PATH}")
+    set(CONFIG_CFLAGS "-I${CONFIG_PREFIX}/${GDAL_INSTALL_INCLUDEDIR}")
+
+    get_property(target_lib_name TARGET "${arg_TARGET}" PROPERTY OUTPUT_NAME)
+    set(CONFIG_LIBS "${CMAKE_LINK_LIBRARY_FLAG}${target_lib_name}")
+    if(NOT CONFIG_PREFIX IN_LIST CMAKE_C_IMPLICIT_LINK_LIBRARIES)
+        string(PREPEND CONFIG_LIBS "${CMAKE_LIBRARY_PATH_FLAG}${CONFIG_PREFIX}/${CMAKE_INSTALL_LIBDIR} ")
+    endif()
+
+    get_property(libs GLOBAL PROPERTY "${arg_GLOBAL_PROPERTY}")
+    if(NOT MSVC AND CMAKE_THREAD_LIBS_INIT)
+        list(APPEND libs ${CMAKE_THREAD_LIBS_INIT})
+    endif()
+    list(APPEND libs ${CMAKE_CXX_IMPLICIT_LINK_LIBRARIES})
+    list(REMOVE_ITEM libs ${CMAKE_C_IMPLICIT_LINK_LIBRARIES})
+    gdal_get_lflags(CONFIG_DEP_LIBS ${libs})
+    if(NOT BUILD_SHARED_LIBS)
+        # Make `--libs` simply work, even for static builds.
+        string(APPEND CONFIG_LIBS " ${CONFIG_DEP_LIBS}")
+        set(CONFIG_DEP_LIBS "")
```

Comment:
This changes the behaviour for `gdal-config` and `pkg-config` for static linkage: The user only needs to use `--libs` regardles of GDAL linkage. For static linkage, `gdal-config --dep-libs` returns an empty lib.